mlx/fast.h

```diff
@@ -61,4 +63,54 @@ array affine_dequantize(
     int bits = 4,
     StreamOrDevice s = {});
 
+typedef std::variant<int, bool, Dtype> TemplateArg;
+
+class MetalKernel {
+ public:
+  MetalKernel(
+      const std::string& name,
+      const std::string& source,
+      std::map<std::string, std::vector<int>> output_shapes,
+      std::map<std::string, Dtype> output_dtypes,
+      std::tuple<int, int, int> grid,
+      std::tuple<int, int, int> threadgroup,
+      bool ensure_row_contiguous,
+      bool verbose)
+      : name_(name),
+        source_(source),
+        output_shapes_(output_shapes),
+        output_dtypes_(output_dtypes),
+        grid_(grid),
+        threadgroup_(threadgroup),
+        ensure_row_contiguous_(ensure_row_contiguous),
+        verbose_(verbose) {
+    validate_output_shapes();
+  }
+
+  std::map<std::string, array> run(
+      std::map<std::string, array>& inputs,
```

Comment:
Although I see why this would be so much nicer and I am not sure we should change it I have to point out that it deviates from the rest of the C++ API which takes vectors of arrays. Also I would change that to `unordered_map`.